keps/sig-architecture/2527-clarify-status-observations-vs-rbac/README.md

```diff
@@ -0,0 +1,697 @@
+
+### Non-Goals
+
+* Retroactively apply this pattern to change existing GA APIs.
+* Necessarily apply this pattern to change pre-GA APIs (though they may choose
+  to follow it).
+* Provide arbitrary storage for controller state.
+* Provide distinct access control for subsets of spec or status.
+
+## Proposal
+
+This KEP proposes to:
+
+1) Get rid of the [if it got lost](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status)
+litmus test for status fields
+2) Acknowledge that spec/status are a useful and meaningful split for access control
+3) Document one or more patterns for APIs that need a trusted acknowledgement
+   of the spec
+
+Depending on feedback, there may be other approaches to solving the same
+problems.
+
+### Risks and Mitigations
+
+The current guidance exists for a reason.  It does encourage a sort of
+idealogical cleanliness.  However, it is not universally adhered to and it
+neglects the reality of these request/acknowledge APIs.  On the whole, this KEP
+posits that the current guidance is a net negative.
+
+## Design Details
+
+<<[UNRESOLVED ideas to debate]>>
+
+### Option 1: Loosen status
+
+Remove the idea that status fields _must_ be from observation.  Allow controllers
+to write values to status that represent allocations or acknowledged requests.
+Document that status fields are best when they represent observed state, but do
+not _require_ it.
+
+This has [already happened](https://github.com/kubernetes/enhancements/pull/2308/files#r567809465)
+at least once.
+
+#### Examples
+
+Some of these are variants of existing APIs and some are hypothetical.
+
+1) User creates a Service and sets `Service.spec.type = LoadBalancer`.  The
+cloud controller sets `Service.status.loadBalancer.ingress.ip` in response.  If
+the user sets `Service.spec.loadBalancerIP` to a specific value, the cloud
+controller will either successfully use that IP and and set it as before (ACK),
+or it will not set any value in `Service.status.loadBalancer.ingress` (NAK).
+
+2) Given a Pod, user patches `Pod.spec.containers[0].resources.requests[cpu]`
+to a new value.  Kubelet sees this as a request and, if viable, sets
+`Pod.status.containerStatuses[0].resources.requests[cpu]` to the same value.
+
+3) User creates a `Bucket`, setting `Bucket.spec.instance = artifacts-prod`.
+The bucket controller verifies that the namespace is allowed to use the
+artifacts-prod bucket and, if so, sets `Bucket.status.instance` to the same
+value.
+
+#### Tradeoffs
+
+Pro: Easy and compatible with existing uses.
+
+Con: Dilutes the meaning of status.
+
+### Option 2: Add a new top-level stanza to spec/status resources
+
+Keep and strengthen the idea that status fields must be from observation.
+Segregate controller-owned fields to a new stanza, parallel to `spec` and
+`status`, which can be RBAC'ed explicitly.  For the sake of this doc, let's
+call it `control`.
```

Comment:
I'd be much more open to something like giving N controllers each their own "status" stanza.

Reply:
I like this too, or alternatively a hybrid of observable and non-observable sub-keys in status to provide clarity which fields can be reconstructed from observation and which fields carry (potentially) non-recoverable state.